### DISCUSSION TITLE
Variante de ejercicio 11 de Vol. 2

### DISCUSSION THREAD

**Discussion by santiemanuel** [santiemanuel/milProgramadoresQA#9] [Preguntas y Respuestas] [answered]
Un programa codifica un mensaje de la siguiente manera

1. Se ingresa una palabra inicial.
2. Se ingresa un número de la misma longitud que la palabra.
3. Para cada letra de la palabra original, se agregan X caracteres aleatorios después de esta.

Ejemplo:
Palabra: CODE
Código: 4161

Salidas posibles:
- **C**WZJL**O**J**D**FCDLOW**E**P
- CMSSHOZDSQSJUPEG
- CBDKWOZDVVFZQXEA

Completado el algoritmo de cifrado, desarrolle el algoritmo que dada una palabra cifrada y el código, devuelva la palabra original.

**Comment by ailednitram** [comment]
Hola! Aquí dejo una posible solución. Saludos!

https://colab.research.google.com/drive/1QNWKdv6lyAt8mCigsmhh9P3uVD-2EpBK#scrollTo=5zZVNZ7NAhLC

**Reply by santiemanuel** [reply]
Buenas, el Colab pide permisos de acceso, podrías pegarlo al código con esta sintaxis? Es el caracter ` tres veces antes de python y en una línea nueva después del código.
![image](https://user-images.githubusercontent.com/3680835/190872050-460b0136-cba8-4950-ac91-6825c336cde6.png)

**Reply by ailednitram** [reply]
```python
cifrada=list('PHDKCHSUYDJCTHDHJCNCUAONJFLNCKF')
aux=[]
code=732643
length=len(str(code))
for i in range(length):
    a=code%10+1
    code=code//10
    aux.append(cifrada[-a])
    cifrada=cifrada[0:-a]
aux.reverse()
print(aux)
```

**Reply by santiemanuel** [reply]
Bien, hace el proceso de descifrar el mensaje. Pensaste en el cifrado? Para que te devuelva esa lista que creaste.
Y otra cosa, el código lo tenés como un `int`, si yo quiero que después de la primera letra hayan 0 caracteres random no se podría.
Tipo:
```
CODE
0122
COQDMVEOZ
```

**Reply by ailednitram** [reply]
Sí, probé descifrando varias opciones y funciona. El cero es un problema, sí, ahí lo estoy viendo.

**Reply by ailednitram** [reply]
```python
cifrada=list('HOJLSDAOR')
aux=[]
code='0122'
altCode=int(code)
length=len(str(code))
for i in range(length):
    a=altCode%10+1
    
    altCode=altCode//10
    aux.append(cifrada[-a])
    cifrada=cifrada[0:-a]
aux.reverse()
print(aux)
```

**Reply by santiemanuel** [reply]
Listo ahora sí, lo que veo que planteaste es recorrer la cadena cifrada al revés y mover X posiciones hacia atrás para llegar a cada caracter original de la palabra, así lo pensaste?

**Reply by ailednitram** [reply]
Tal cual. Al principio lo pensé en orden pero se me complicaba al ir descomponiendo el número.. me pareció más fácil así

**Comment by santiemanuel** [comment, marked as answer]
Dejo la solución que planteé yo para este ejercicio.

# Parte de cifrado:

```py
from random import randint

alfabeto = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

palabra = "CODE"
codigo =  "1512"
codificada = ""

for (i, letra) in enumerate(palabra):
    codificada += letra
    for j in range(int(codigo[i])):
        codificada += alfabeto[randint(0,25)]

print(codificada)
```

El alfabeto se lo podría sacar con la biblioteca string, lo puse para ahorrar un import.

Luego tenemos la palabra y el código como cadenas de misma longitud, seguido de una cadena vacía en donde vamos a guardar la palabra codificada.

En el ciclo, uso la función enumerate, esta nos crea una secuencia numérica en la primera variable del for.
Entonces, letra vale C, O, D, E, mientras que i vale 0, 1, 2, 3. Esto lo uso para tener un índice para posicionarme en un elemento del código.

El ciclo más interior solo lo uso para iterar `digito` veces para agregar esa cantidad de letras aleatorias a mi cadena codificada.

Una vez terminado de agregar estos caracteres, paso al siguiente dígito del código y repito el proceso.

Terminados los dos ciclos obtengo mi palabra codificada, como por ejemplo (porque tiene caracteres aleatorios):
```
CIOIUCFDDEEDU
```

# Parte de descifrado:

Esto se puede resolver tanto con ciclos while como con ciclos for. Muestro las dos alternativas.

## Ciclo While

Tenemos nuestra cadena `codificada=CIOIUCFDDEEDU`

```py
longitud=len(codificada)
i = 0
id_cod = 0
decodificada = ""
while i < longitud:
    decodificada += codificada[i]
    i += int(codigo[id_cod]) + 1
    id_cod += 1

print(decodificada)
```

Primero averiguo la longitud de esta en longitud, y creo dos índices:
- `i` para avanzar el posiciones de la cadena codificada
- `id_cod` para avanzar por los dígitos del código

Como longitud es de la palabra codificada la uso como condición del ciclo usando su índice i para recorrer toda la cadena.

Ahora empiezo agregando la primera letra directamente dado que el primer caracter nunca se mueve en el mensaje codificado.

Luego, avanzo el valor de i en el valor del dígito para esa posición y le sumo 1. El sumar 1 es para pasar al siguiente caracter, si el código dice por ejemplo 0012, el código no avanzará el índice i y entonces el ciclo no terminará nunca.

Una vez avanzado las "dígito del código"-posiciones avanzo en 1 la posición del índice del código para repetir el proceso dígito a dígito hasta obtener el mensaje original.

## Ciclo For

```py
i = 0
decodificada = ""

for digito in codigo:
    decodificada += codificada[i]
    i += int(digito) + 1

print(decodificada)
```

Aquí tenemos menos variables por la facilidad de este ciclo, i lo usamos como índice de un caracter en el mensaje codificado.

El ciclo for en su primera línea toma un dígito de la cadena código y lo guarda en la variable 'digito'.

Como vimos antes, la posición 0 es correcta así que agregamos esta letra inicialmente.

Ahora, para avanzar el índice de la palabra codificada usamos el dígito que nos provee el for y le aumentamos 1 por la misma razón en la resolución del while. De esta forma tenemos el mismo proceso con una sintaxis más sencilla de interpretar.

Si queda alguna duda me responden a este mensaje.
